Add interior cabin rudder trim system

wlbragg opened the pull request:
Fixes #1217 

Ready for testing.
@gilbertohasnofb  I think it still needs your handy work on the texture, this one was just for testing and as a template.
@legoboyvdlp I would love it if you would take a good look at the FDM portion and test it and make any modifications necessary. I didn't mean to do the FDM portion, I was just trying to get to a point where I knew everything was working together correctly.

Also @gilbertohasnofb I want your input on the proposed menu changes as well as the current menu changes.

:warning::warning::warning: DO NOT MERGE :warning::warning::warning:

legoboyvdlp commented:
Yep, the cockpit trim should be reset to 0 - other than that, everything looks good. Maybe the lever could have some sound, maybe the same sound as the primer or alternate static?

wlbragg commented:
How do you get the sound to work on this kind of animation? I thought the primer-out sound would be perfect for it but I can't get it to work?
I must be missing something. 
And I couldn't even find the code that is activating the sound of the elevator dial. It looks like it should be "dial-trim-wheel" in the sound file, but I don't see anywhere where it is actioned?
This didn't produce sound for some reason?
~~~
<animation>
        <type>knob</type>
        <object-name>RudderTrimKnob</object-name>
        <property>controls/flight/rudder-trim-knob</property>
        <factor>40</factor>
        <drag-direction>horizontal</drag-direction>
        <offset-deg>0</offset-deg>
        <center>
            <x-m>-0.521  </x-m>
            <y-m> 0.00061</y-m>
            <z-m>-0.56081</z-m>
        </center>
        <axis>
            <x>0</x>
            <y>0</y>
            <z>1</z>
        </axis>
        <action>
            <binding>
                <command>property-adjust</command>
                <property>controls/flight/rudder-trim-knob</property>
                <factor>0.01</factor>
                <min>-.18</min>
                <max>.18</max>
                <wrap>false</wrap>
            </binding>
            <binding>
                <command>nasal</command>
                <script>c172p.click("primer-out", 1.2)</script>
            </binding>
        </action>
        <hovered>
            <binding>
                <command>set-tooltip</command>
                <label>Rudder trim: %3.2f</label>
                <tooltip-id>rudder-trim</tooltip-id>
                <property>controls/flight/rudder-trim-knob</property>
            </binding>
        </hovered>
    </animation>
~~~

See my last message at https://github.com/c172p-team/c172p/issues/1217 for an explanation of the persistent properties of the rudder trim system.

wlbragg commented:
@gilbertohasnofb @legoboyvdlp any idea why I cant get this sound to work?

legoboyvdlp commented:
Not really, I'm afraid.

gilbertohasnofb commented:
Your last commit has broken the texturing for 4X-CHV for some reason. The rudder trim should look white for all liveries other than that one (since their panel.png file have not been updated), but 4X-CHV should look fine.

gilbertohasnofb commented:
@wlbragg I decided to sync all panel textures for the rudder trim stuff and I have pushed it already, so that you can test the livery system when implementing it. I hope that works well for you.

wlbragg commented:
@gilbertohasnofb Sorry about the merge, I committed my local work before I realized you pushed something and didn't try to rebase my work on top of yours.
All is good though.
I had to move the rudder trim objects from their own stand alone .ac to the c172p.ac. The way we are trying to do a standalone .ac using a texture from a different directory and also trying to do an effect on top of it just wasn't working. This is the only object we were trying to do this with and I couldn't get the light map to work on it with the texture being in panel.png. Everything I looked at told me it should work, but it didn't. After trying it a couple different ways it still didn't work.
So I added the objects to c172p.blend and simply use a select animation in the GUI to control it's visibility (select animation is the same as it was before). That put the effect definition in the c172p.xml file and fixed the LM issue.
I checked a couple liveries and they appear to work correctly.
So as far as I am concerned this is finished and can be merged after @legoboyvdlp  @dany93 and yourself test and verify the effectiveness of the trim settings.

wlbragg commented:
>Could you please map only the bottom half of the metal texture to the lever?

Yep, will do and push shortly.

gilbertohasnofb commented:
Looks good to me! @legoboyvdlp would you like to have a last look to see if you spot anything?

> Sorry about the merge, I committed my local work before I realized you pushed something and didn't try to rebase my work on top of yours.

No worries about that, that's bound to happen when two people work on a same branch. I personally don't think this is an issue. If you do spot it on your side before pushing it then you can always try to rebase it locally (perhaps backup your work if you are neurotic about losing it). But I don't think it's a big deal.

legoboyvdlp commented:
Feedback:

On my 15" screen, the aircraft options dialog is too tall:

![image](https://user-images.githubusercontent.com/10434911/47522675-1a1f1380-d88e-11e8-84cd-6dc73bdb4838.png)


I see this on the rudder trim branch?!?!?!

![image](https://user-images.githubusercontent.com/10434911/47522686-260ad580-d88e-11e8-93de-55083cf371cc.png)

legoboyvdlp commented:
The rudder trim effect is barely noticeable. Also, when I put full left, the skip-slid is telling me to apply more left rudder, when I put full right it tells me to put a slight bit more right. Finally, the HUD control does not move with the rudder trim?

legoboyvdlp commented:
I really hope that bug with the frost is fixed by the frost-and-fog PR!

legoboyvdlp commented:
Yep, that would be great.

dany93 commented:
> The rudder trim effect is barely noticeable. Also, when I put full left, the skip-slid is telling me to apply more left rudder, when I put full right it tells me to put a slight bit more right

True.
- The effect is weak,
- It is in the wrong direction (rudder trim 'more left' must make the aircraft yaw and turn leftwards). The slip ball must go rightwards. Like left rudder would do.

wlbragg commented:
OK, what needs to be changed and by how much? @legoboyvdlp if you know what needs to be done here I have no problem if you take over and fix it.

wlbragg commented:
>I see this on the rudder trim branch at sunset and very cold temperatures?

Would have been a good test to switch to the fog/frost branch and see if you could recreate for a comparison.

Just so I am clear, other than the artistic aspect of the ice texture, is there something else you are noticing that I am not seeing? Specifically is it the streaking on the left window that is the worst part in your opinion?

legoboyvdlp commented on c172p.xml:
```diff
@@ -810,9 +810,23 @@
+                    <table>
+                        <independentVar lookup="row">/controls/flight/rudder-trim-knob</independentVar>
+                        <tableData>
+                            -.18   -0.05
```
```suggestion
                           -.18    0.10
                            0.00    0.00
                             .18   -0.10
```

legoboyvdlp commented:
I pushed a commit which fixes the reversed rudder trim. Don't ask me why what would seem the right direction isn't 😩 

I also increased the effect to +- 1. This can be adjusted more if desired.

wlbragg commented:
@legoboyvdlp thanks for handling this.

@legoboyvdlp  @gilbertohasnofb I can't believe I did something that could be very dangerous, but in this case will probably be OK. 
@gilbertohasnofb after we decided to use the panel .png texture for this and subsequently I added the mesh to c172p.blend and c172p.ac I did what we have been desperately trying to avoid , and that is to have two blend file modifications of the same blend going at the same time. I stepped back for a day and when I cam back to it I just now noticed i pushed the blend and .ac here. Same blend and .ac I am pushing to the fog/frost branch. The latest fog/frost branch blend does not have the rudder trim mesh yet.
Good news is I realized it before any damage is done and it is a matter of coordination to keep it working correctly.
My suggested solution is to push the same blend to both branches as of right now and then from here on out make sure I keep them synchronized until one of these branches is merged.
I can't believe it slipped my mind when I reverted to putting the rudder trim mesh into the c172p.blend.
@gilbertohasnofb I think it is important to know why this happened and it was because of the change in strategy of dealing with the texture in panel.png VS the stand alone one I originally started out on. I understand why we decided to switch strategies mid stream like that, and I agreed with the change, but we both need to be really conscious of the ramification when were dealing with these blends.
This is a continuing issue we keep trying to build safeguards for. That is why I always gravitate towards doing so many different issues in the same PR. I took to heart what you suggested and tried to do the rudder trim in it's own PR including the mesh work. That obviously changed midstream and just about bit us.
My only purpose in this post is to point out what happened and why so we can all avoid it in the future.
For now, for this immediate issue, we simply need to coordinate merges and make sure the most current blend is in the last PR merged.
I apologize for yet another mistake in procedure, but it is so easy to forget this simple concept when you proceed down one path and switch mid stream. My mind had long since been programed to know these were separate blends until they weren't any longer and I simply missed it.

gilbertohasnofb commented:
@wlbragg I personally feel that a much safer solution is to only do one 3D work at a time and wait for that PR to be merged before moving on to something else. In this case, for instance, there was no rush to create the rudder trim. I understand that this limits the number of parallel things we can do, but it would definitely avoid these problems. Also, smaller PRs == smaller chances of headaches, faster merging, higher rates of features hitting the `master`, etc.

> I think it is important to know why this happened and it was because of the change in strategy of dealing with the texture in panel.png VS the stand alone one I originally started out on. I understand why we decided to switch strategies mid stream like that, and I agreed with the change, but we both need to be really conscious of the ramification when were dealing with these blends.

I am not sure I understand what you mean, so please correct me if I am misunderstanding you: Often we need to modify the UV-mapping after a texture is created, so I don't think it would be a good approach to model something on branch A, add a texture on branch B and just hope it will look amazing. The back and forth we have is fundamental to a high quality work. I think we must always be able to adjust the model and UV-mapping when creating new stuff.

dany93 commented:
@legoboyvdlp wrote
> ..... the reversed rudder trim. Don't ask me why what would seem the right direction isn't

I think that is because the Z axis is downwards.
Right rudder control gives a positive value (`controls/flight/rudder`),
The rudder rotates rightwards (counter clockwise), which gives a negative angle (`fdm/jsbsim/fcs/rudder-pos-norm` negative) with Z as the reference.

The elevator does not have this sign problem: pitch down command is positive, the elevator surface rotates downwards, which gives a positive angle with the Y axis to the right side.

(please be critical, to be confirmed)

legoboyvdlp commented:
Ah ok - that makes sense!
Thanks 😃

wlbragg commented:
>I personally feel that a much safer solution is to only do one 3D work at a time and wait for that PR to be merged before moving on to something else. In this case, for instance, there was no rush to create the rudder trim. I understand that this limits the number of parallel things we can do, but it would definitely avoid these problems. Also, smaller PRs == smaller chances of headaches, faster merging, higher rates of features hitting the master, etc.

I can't disagree with any of this even though I like the freedom to roll when the urge strikes and time permits. There is no reason I can't do that locally still but limit the Git-Hub work to one issue at a time. 

>I am not sure I understand what you mean, so please correct me if I am misunderstanding you: 

Just that I started all this work in it's own encapsulation and it diverged to the other branch. I took to heart your suggestion of doing it in its own blend which is how it started, its own separate blend and texture. It changed because that created undo and extra work for you in the texturing department (liveries). So mid stream we change it to use the panel.png texture which subsequently changed the effects requirement need for the mesh to be in the c172p.blend VS the original stand alone blend I started out with.
That's all, and then I overlooked the ramification of having the same blend now working in two different PR's because in my mind I had already sorted that out by using two stand alone blends.
Really nothing we can do about it, I was just letting you know why this happened this way. Your requirement (or request) for adding the texture to panel.png put the ball in motion to put the mesh there as well. That provided the mechanics for me to mess up by switching mesh from a standalone blend to a blend already in use in a different PR.

gilbertohasnofb commented:
@wlbragg Let's perhaps try to not overthink this, we have been working great for many years and all our problems so far were but minor hiccups at best. Please do work in the way it suits you best and I am sure we will be fine with that. Perhaps we can try to keep these things in mind and try to merge PRs as soon as they are ready, but if a new project comes up and you feel like getting started with it while the other one isn't done yet, hey, that's better than if no one would be working on either of them. My apologies if I dragged things too long and if we have been exchanging too many messages about the mechanics of git, let's just try to move on and get back developing a plane, which is much more fun for all of us :smile:

gilbertohasnofb commented:
@wlbragg what is the status of this PR, are we ready to merge?

wlbragg commented:
I think we need @dany93 to sign off on this. I think strength of the trim settings is all that was left to finalize.

dany93 commented:
I don't know more than you and lego about the efficiency of this rudder-trim setting in RL.
My opinion is that it is a bit weak. An amplitude of +/- 0.20 (instead of +/- 0.10) gives a shift of the slip ball by about one ball diameter at full effect, which still lets a correct accuracy.

I never felt the need for it with the FG c172p. It might be used during the climbing leg but I don't think it is the case for such a temporary stage. All the more that it can be in a airport circuit where you have to change at each leg.
I have this rudder-trim setting available on my Joystick. I only used it for the P51D, in a step of tests and adjustments.

It was @legoboyvdlp's idea, I guess he felt the need for it. The best would be to test its efficiency in conditions where you need it.

wlbragg commented:
@dany93 thanks for your impressions. I set it to +- .20. @legoboyvdlp if you think this is sufficient then this branch should be ready to merge.

legoboyvdlp commented:
That's definitely fine. I'll go ahead and merge this then :smiley: 

Thanks for making this!

wlbragg commented:
Your welcome and I like having the option when flying cross country with a cross wind. I think it was a good option to add.

dany93 commented:
> I like having the option when flying cross country with a cross wind

No. With a cross wind you compensate it by flying 'in crab' **in relation to the ground** but the coordinated flight is exactly the same: the apparent wind must come from the front, symmetrically. The rudder action to keep the slip ball centered is exactly the same. Same position for the same conditions.
The main difference is when you observe the ground. Not the airflow.

A large difference is at landing with a cross wind, after decrabbing.

wlbragg commented:
>If you cannot see the ground, you cannot know.

Good point!

>A large difference is at landing with a cross wind, after decrabbing.

Makes sense, thanks.
